lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -5667,6 +5532,75 @@ Value *PatchInOutImportExport::getDeviceIndex(Instruction *insertPos) {
   }
 }
 
+// =====================================================================================================================
+// Records export info of vertex attributes
+//
+// @param location : Vertex attribute location
+// @param attribValues : Values of this vertex attribute to export
+void PatchInOutImportExport::recordVertexAttribExport(unsigned location, ArrayRef<Value *> attribValues) {
+  assert(m_shaderStage == ShaderStageVertex || m_shaderStage == ShaderStageTessEval ||
+         m_shaderStage == ShaderStageCopyShader); // Valid shader stages
+  assert(attribValues.size() == 4);               // Must have 4 elements, corresponds to <4 x float>
+
+  auto undef = UndefValue::get(Type::getFloatTy(*m_context));
+
+  // Vertex attribute not existing, insert a new one and initialize it
+  if (m_attribExports.find(location) == m_attribExports.end()) {
+    for (unsigned i = 0; i < 4; ++i)
+      m_attribExports[location][i] = undef;
+  }
+
+  for (unsigned i = 0; i < 4; ++i) {
+    assert(attribValues[i]);
+    if (!isa<UndefValue>(attribValues[i])) {
+      assert(isa<UndefValue>(m_attribExports[location][i])); // Must be invalid, couldn't have overlapping
```

Comment:
I can't parse this comment.

Reply:
This means if the recorded values are undef and we are going to record new values, this case is valid. See this case:

Valid `attrib0: recorded value <undef, undef, 1.0, 2.0>, new value <3.0, 4.0, undef, undef>`
Invalid `attrib0: recorded value <undef, undef, 1.0, 2.0>, new value <3.0, 4.0, 5.0, undef>`

In the invalid case, 5.0 will overwrite 1.0. This is invalid. The GLSL spec disallows such overlapping cases. GLSLang should report a error.

Reply:
Ah, I see. Can you turn this comment into an assertion message, maybe something like: `"New value for a non-undef recorded value must be an undef" and add your example as a comment?

Reply:
Sure, will done.